[Feature] Block user [Fix] current chat in bold on Linux [Fix] Username is bold on Linux

yenda opened the pull request:
fix #6033 remove logic from view
fix #1660 block user

- usernames in bold font on Linux
- prevent use of rtl char in username to impersonate
other other user random name
- current chat in bold font
- remove defview and subscriptions to use pure functions with params
- allow user to block/unblock a contact
- this will remove blocked user from chat list and
hid any message sent on any channel by blocked user

# Step to test

Plateform: desktop

sanity check
- test that chats, public chats and private chats are working correctly, responses are displayed properly
- test that users can be blocked in their profile

status: ready

Commit:
[refactor] chat views on desktop

- usernames in bold font on Linux
- remove defview and subscriptions to use pure functions with params
- prevent use of rtl char in username to impersonate
other other user random name

ghost commented:
<!--prchecklist--> 
### Pull Request Checklist
- [x] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://docs.status.im/docs/build_status.html))?

flexsurfer commented:
@yenda could you please provide some description?

dmitryn commented:
@yenda please rebase and squash commits

yenda commented:
@flexsurfer done
@dmitryn done

dmitryn commented on src/status_im/ui/screens/desktop/main/chat/views.cljs:
```diff
@@ -205,128 +186,119 @@
-       [send-button inp-ref network-status]])))
+(defn chat-text-input [chat-id input-text]
+  (let [inp-ref (atom nil)
+        network-status [:network-status]]
```
shouldn't  `network-status` be reagent subscription? now it looks like as const vector `[:network-status]`

churik commented:
@yenda 
New issue appeared, reproducible only in PR build.
Chat view is broken if you send several messages to chat with last message from yesterday or if you set custom date in the past in your settings (whole or part of chat area is blank)

Possible to reproduce if you have chats with last messages from yesterday or if you set custom date on your laptop to the past.

To emulate it from scratch:
- set your Mac OSx date to past (i.e. yesterday)
- create public channel
- try to send message to this channel

Expected: can see message
Actual: can see only blank screen, no new messages are shown in chat area
Video: http://take.ms/0bpgs
Logs: [Status.log](https://github.com/status-im/status-react/files/2565736/Status.log), [geth.log](https://github.com/status-im/status-react/files/2565738/geth.log)


<blockquote><img src="https://api.monosnap.com/rpc/file/download?id=HhZzlXt9KBJmjtYUI8TWyBhq2wsD1W&type=preview" width="48" align="right"><div>Monosnap screenshot tool</div><div><strong><a href="https://monosnap.com/file/HhZzlXt9KBJmjtYUI8TWyBhq2wsD1W">File "screencast 2018-11-09 12-57-54.mp4"</a></strong></div><div>Monosnap — the best tool to take & share your screenshots.</div></blockquote>

yenda commented:
@churik I fixed it, I think an easiest way to reproduce was to open a new public chat and then switch to already open ones, at least for me it was like this.

This PR now introduces the following glitch when replying to a chat:

![screenshot from 2018-11-09 15-44-47](https://user-images.githubusercontent.com/1181225/48269006-cfc59700-e436-11e8-92e7-bbdf71871934.png)

Personally I don't think this is blocking, and it should make fixing flat-list a priority wdyt ?

churik commented:
@yenda 
I believe it should be fixed before merge.
Thanks for creating corresponding issue on RN-side!
Blocked by https://github.com/status-im/react-native-desktop/issues/391

Commit:
[refactor] chat list on desktop

- current chat in bold font
- remove defview and subscriptions to use pure functions with params

Commit:
[feature] add block user feature

- allow user to block/unblock a contact
- this will remove blocked user from chat list and
hid any message sent on any channel by blocked user

cammellos commented on src/status_im/chat/subs.cljs:
```diff
@@ -98,30 +111,96 @@
+ :<- [:get-chats]
+ :<- [:contacts/blocked]
+ :<- [:account/account]
+ (fn [[contacts chats blocked-contacts {:keys [dev-mode?]}]]
```
Please add some tests covering this (at least the newly written code).

yenda replied:
I was planning to come back to this whole file for a cleanup when doing the feature for mobile but since it's blocked by this flatlist bug I might include the mobile as well

cammellos commented on src/status_im/contact/subs.cljs:
```diff
@@ -11,11 +11,20 @@
 
+(reg-sub :contacts/blocked
+         :<- [:get-contacts]
+         (fn [contacts]
```
extract and test?

cammellos reviewed (changes requested):
Please add some tests to cover new functionalities

Commit:
[refactor] chat views and chat list on mobile

- apply refactoring of chats to mobile
- simplify subscriptions accordingly

Commit:
[refactor] message delivery status

- move computations from views to subscriptions

Commit:
[fix] align min input container height with tab menu

Commit:
[fixup] groupchats

dmitryn commented on src/status_im/ui/screens/chat/message/message.cljs:
```diff
@@ -41,22 +37,21 @@
-; so we pad the invisible placeholder with some spaces to avoid having too
-; close to the text.
+                                        ; We can't use CSS as nested Text element don't accept margins nor padding
+                                        ; so we pad the invisible placeholder with some spaces to avoid having too
```
looks like too much indent here

cammellos commented on src/status_im/ui/screens/desktop/main/chat/views.cljs:
```diff
@@ -68,298 +64,247 @@
+          {:style (styles/chat-profile-icon colors/black)}]]
+        [react/text {:style (styles/contact-card-text colors/red)}
+         (if blocked?
+           "Unblock user"
```
translations

cammellos reviewed (approved):
Too big to really understand what has changed or follow the logic, please smaller PRs :) I think for plain refactoring small prs with `skip-qa` is ok , I went line by line to see if there's anything obviously odd, but looks ok to me. Please test as well pairing & group chats on a rebased version. 
5000 SNT on my nameday if no tests :)

yenda commented:
@camellos yes 5000SNT by Nov 30th if I didn't add new tests by then
Regarding the length it's already cut into smaller commits, the issue was that the first commit on which it is built was blocked by a UI glitch in react-native-desktop

cammellos commented:
@yenda my bad, didn't see that, thanks for pointing out

churik commented:
Testing is still WIP, comment will be updated:

Mac OSx:
###  1. Last row is cut in chat list if message contains 1+ rows*
Screen:
<img width="568" alt="macos" src="https://user-images.githubusercontent.com/4557972/48496261-f23e2280-e831-11e8-9f6f-91358173e93e.png">

### 2. Space between last message and previous ones are different
<img width="568" alt="hsdf" src="https://user-images.githubusercontent.com/4557972/48496793-1817f700-e833-11e8-8243-e7739e2e0f84.png">
The space between last message and text input area is also depending on message is incoming or outgoing:
<img width="568" alt="outg" src="https://user-images.githubusercontent.com/4557972/48498642-f6b90a00-e836-11e8-8687-617e68f116a0.png">
<img width="568" alt="inc" src="https://user-images.githubusercontent.com/4557972/48498644-f751a080-e836-11e8-8dd0-59c45269d077.png">

### 3. No "Add to contacts" in 1-1 chat when someone adds you as contact
Steps:
1) log in to desktop app as user A
2) tap on Share profile code
3) log in as another user on mobile
4) start new chat > scan QR
5) send message to user A

**Expected result: **

User A can see "Add to contacts " in 1-1 chat

**Actual result: **

<img width="568" alt="add_to_contacts" src="https://user-images.githubusercontent.com/4557972/48499070-ddfd2400-e837-11e8-8b5e-9ae3ac126653.png">

### 4. Spam window doesn't block user from sending new messages 
Steps:
1) log in to desktop app as user A
2) join any public chat
3) send quickly 5+ messages in a row

**Expected result: **

Can send only 5 messages, then popup appears and user cannot send more until popup will be closed

**Actual result: **
Spam window appears, but user still can send messages 
<img width="568" alt="spam" src="https://user-images.githubusercontent.com/4557972/48499819-94add400-e839-11e8-8340-ccac8da16bdf.png">


### 5. Switching beween chats is slowed down 
Video from current PR:  https://monosnap.com/file/4esSAd79M1NTzRKJy6qdGwhZHXCqvs
Video from nightly: http://take.ms/HWVui

<blockquote><img src="https://api.monosnap.com/rpc/file/download?id=4esSAd79M1NTzRKJy6qdGwhZHXCqvs&type=preview" width="48" align="right"><div>Monosnap screenshot tool</div><div><strong><a href="https://monosnap.com/file/4esSAd79M1NTzRKJy6qdGwhZHXCqvs">File "screencast 2018-11-14 18-12-05.mp4"</a></strong></div><div>Monosnap — the best tool to take & share your screenshots.</div></blockquote>
<blockquote><img src="https://api.monosnap.com/rpc/file/download?id=MakgaGhY8yHzxv7QRZqVtKsFd1RQve&type=preview" width="48" align="right"><div>Monosnap screenshot tool</div><div><strong><a href="https://monosnap.com/file/MakgaGhY8yHzxv7QRZqVtKsFd1RQve">File "screencast 2018-11-14 18-21-50.mp4"</a></strong></div><div>Monosnap — the best tool to take & share your screenshots.</div></blockquote>

yenda commented:
1. seems like an OSX specific problem, I increased the max size because on Linux the bottom of some letters was cut. I changed it back for OSX.
2. sounds like a minor issue could we bounty it ? The second part is same as in develop, it's because of the message status on the bottom right corner
3. fixed
4. it's in develop as well

Commit:
fixup

yenda commented:
5. regarding the slow down it's due to the fact that some stuff is now recalculated, I will work on optimizations later on. It's necessary otherwise you would see preview of blocked messages and counter would include blocked messages the way it was working before

chadyj commented:
Bit more context for blocking >> https://github.com/status-im/status-react/issues/6735

churik commented:
@yenda thank you!
### 2. First part is minor, agree, but second it is not working the same on current nightly and PR.

On current nightly:
<img width="573" alt="in_outg" src="https://user-images.githubusercontent.com/4557972/48542213-df256400-e8be-11e8-8e37-4d6966ace9aa.png">

On PR build:
<img width="560" alt="cur_pr" src="https://user-images.githubusercontent.com/4557972/48548760-0fc0ca00-e8ce-11e8-8c3c-0e01cfc41270.png">
Difference is noticeable.

### 6. `Invalid arity:` 1 when creating account on IOS and Android
After creating account I've got an error:
<img width="322" alt="android_png" src="https://user-images.githubusercontent.com/4557972/48546998-f158cf80-e8c9-11e8-90f4-fe9b1f374234.png">
After restarting app and login I see the same error.
Testing on mobile is blocked.

### 7. When you tap on "Block user" on profile, then click on "Send message", you  will be redirected to empty chat without UI

Video: http://take.ms/UvhXT
Log: [Status.log](https://github.com/status-im/status-react/files/2585002/Status.log)
I you tried to send some messages to this chat, receiver will receive them, but you cannot see them (only empty screen will be shown)
<img width="568" alt="monosnap 2018-11-15 12-31-58" src="https://user-images.githubusercontent.com/4557972/48550255-8f509800-e8d2-11e8-87ca-f56d3ea464ee.png">


### 8. `TypeError: e.call is not a function` when tried to block a user who is not in my contact list
Steps:
- created new account on release build 0.9.30 (User A)
- login to desktop app (User B)
- click on "Share my contact code" (User B)
- added user B as contact (user A)
- block user  from his profile (User B)
<img width="568" alt="error_0" src="https://user-images.githubusercontent.com/4557972/48549681-be660a00-e8d0-11e8-9883-b4982b2695a9.png">
Log: [Status.log](https://github.com/status-im/status-react/files/2585040/Status.log)

### 9. Messages that are coming from blocked user are counted in unread message badge on Mac OSx

This badge is updated when messages from blocked user are coming.
<img width="28" alt="unread_0" src="https://user-images.githubusercontent.com/4557972/48550467-1aca2900-e8d3-11e8-8112-ff0206358d2b.png">

### 10. Should be optimized in terms of performance before merge.
Tests were made on the same account with same chats on different builds.

Action | CPU, % (PR) | Real memory size (PR) | CPU, % (nightly) | Real memory size (nightly) | % increase on Memory Size in PR build
-- | -- | -- | -- | -- | --
Login | 1.79 | 429 | 3 | 385.9 | 11.17
Switching between chats (10 times) | 26.52 | 1010.9 | 27 | 512 | **97.44**
Join chat with 10+ messages | 3 | 591.2 | 11 | 469.2 | 26
Send 10 messages in a row | 1.6 | 772.9 | 1.79 | 500.3 | **54.49**
Switching to Profile and back (5 times) | 27 | 444.4 | 35 | 433.5 | 2.51
Scrolling chat history | 98 | 3190 | 27 | 704.9 | **352.55**
Switching mailserver | 9 | 665.3 | 4 | 626.4 | 6.21




<blockquote><img src="https://api.monosnap.com/rpc/file/download?id=a8tGOVyO2eXILSvRPPrZXYIdFgKLUV&type=preview" width="48" align="right"><div>Monosnap screenshot tool</div><div><strong><a href="https://monosnap.com/file/a8tGOVyO2eXILSvRPPrZXYIdFgKLUV">File "screencast 2018-11-15 12-08-58.mp4"</a></strong></div><div>Monosnap — the best tool to take & share your screenshots.</div></blockquote>

Commit:
updating cljs compiler

Commit:
[fix] improve perfs

- remove message-groups
- make sure chat-list-view remounts when switching chat
to avoid comparing list of messages
- do all message enriching in a single pass to avoid iterating over
message collection multiple times
- separate chat keys from chat.ui keys in app-db

Commit:
fixup

yenda commented:
@churik 

I fixed 7 and 8, can you test performances again they should be better now

Commit:
fixup priority map

rasom commented on src/status_im/chat/db.cljs:
```diff
@@ -0,0 +1,341 @@
+  M1 needs to be displayed before M2
+  so we bucket both in 1999-12-31"
+  [{:keys [acc last-timestamp last-datemark]} {:keys [timestamp datemark] :as msg}]
+  (cond (empty? acc)                                     ; initial element
```
@yenda i believe we agreed once to put comments before the line instead of in the end of line

rasom commented on src/status_im/chat/db.cljs:
```diff
@@ -0,0 +1,341 @@
+             (< timestamp last-timestamp))               ; not out-of-order
+        {:last-timestamp timestamp
+         :last-datemark  datemark
+         :acc            (conj acc {:value last-datemark ; intersperse datemark message
```
80 chars limit

rasom commented on src/status_im/chat/db.cljs:
```diff
@@ -0,0 +1,341 @@
+  [messages blocked-contacts]
+  (remove #(blocked-contacts (:from %)) messages))
+
+(defn- set-previous-message-info
```
function is never used

rasom commented on src/status_im/chat/db.cljs:
```diff
@@ -0,0 +1,341 @@
+(ns status-im.chat.db
+  (:require [clojure.set :as set]
```
namespace is never used

rasom commented on src/status_im/chat/db.cljs:
```diff
@@ -0,0 +1,341 @@
+        (#{:sending :not-sent} status) status
+        command-network-mismatch? :network-mismatch
+        last-outgoing? (if (= message-type :group-user-message)
+                         (get-group-message-delivery-status message (:contacts current-chat) current-public-key)
```
80 chars, please check it in the whole ns, many places where it is exceeded

rasom commented on src/status_im/chat/db.cljs:
```diff
@@ -0,0 +1,341 @@
+                             (str "+ " (- delivery-count 3)))}))
+
+(defn add-metadata
+  [{:keys [from response-to message-id chat-id outgoing]
```
`response-to` is never used

rasom commented on src/status_im/chat/db.cljs:
```diff
@@ -0,0 +1,341 @@
+                   (gfycat/generate-gfy chat-id))))
+
+(defn- enrich-active-chat
+  [{:keys [chat-id messages unviewed-messages group-chat
```
`unviewed-messages` is never used, there are also other unused symbols in the ns, please check them

rasom reviewed (changes requested):
Why is this PR that big? As far as i can see it was approved when its description was much smaller and scope was different. Can it be split into separate PRs? Should we request re-reviews from @flexsurfer  and @dmitryn?

churik commented:
@yenda 
results from https://github.com/status-im/status-react/pull/6652#issuecomment-439285992

Action | CPU, % (PR) | Real memory size (PR) | CPU, % (nightly) | Real memory size (nightly) | % increase on Memory Size in PR build
-- | -- | -- | -- | -- | --
Login | 1.8 | 842 | 3 | 385.9 | **118.2**
Switching between chats (10 times) | 39 | 753.5 | 27 | 512 | 47.2
Join chat with 10+ messages | 1.3 | 598.9 | 11 | 469.2 | 27.6
Send 10 messages in a row | 16 | 734 | 1.79 | 500.3 | 46.7
Switching to Profile and back (5 times) | 34 | 490 | 35 | 433.5 | 13
Scrolling chat history | 91 | 2520 | 27 | 704.9 | **257.5**
Switching mailserver | 8 | 530 | 4 | 626.4 | -15.4


Also I noticed that message ordering is broken now, so if you send several message to channel, it is shown in the middle of the history or overlapping with other messages.

<img width="328" alt="2_order" src="https://user-images.githubusercontent.com/4557972/48624865-dfe8f380-e9ad-11e8-8262-d4e15f514689.png">
<img width="335" alt="1_order" src="https://user-images.githubusercontent.com/4557972/48624866-dfe8f380-e9ad-11e8-91e5-247990c86278.png">

yenda commented:
closed as refactoring was moved to https://github.com/status-im/status-react/pull/6764 and block feature will be in a separate PR
